cmd/provision/main.go

```diff
@@ -98,74 +114,130 @@ func main() {
-		mfCertsURL:       mainflux.Env(envMfCertsURL, defMfCertsURL),
-		Config: provision.Config{
-			MFEmail:          mainflux.Env(envMfUser, defMfUser),
-			MFPass:           mainflux.Env(envMfPass, defMfPass),
-			PredefinedThings: strings.Split(mainflux.Env(envThingIDs, defThingIDs), ","),
-			X509Provision:    provisionX509,
-			BSProvision:      provisionBS,
-			AutoWhiteList:    autoWhiteList,
-			BSContent:        fmt.Sprintf(mainflux.Env(envBSContent, defBSContent), mainflux.Env(envMQTTURL, defMQTTURL)),
+	cfg := provision.Config{
+		Server: provision.ServiceConf{
+			LogLevel:       mainflux.Env(envLogLevel, defLogLevel),
+			CACerts:        mainflux.Env(envCACerts, defCACerts),
+			ServerCert:     mainflux.Env(envServerCert, defServerCert),
+			ServerKey:      mainflux.Env(envServerKey, defServerKey),
+			HTTPPort:       mainflux.Env(envHTTPPort, defHTTPPort),
+			MfBSURL:        mainflux.Env(envMfBSURL, defMfBSURL),
+			MfWhiteListURL: mainflux.Env(envMfBSWhiteListURL, defMfWhiteListURL),
+			MfCertsURL:     mainflux.Env(envMfCertsURL, defMfCertsURL),
+			MfUser:         mainflux.Env(envMfUser, defMfUser),
+			MfPass:         mainflux.Env(envMfPass, defMfPass),
+			MfApiKey:       mainflux.Env(envMfApiKey, defMfApiKey),
+			ThingsLocation: mainflux.Env(envThingsLocation, defThingsLocation),
+			UsersLocation:  mainflux.Env(envUsersLocation, defUsersLocation),
+			TLS:            tls,
+		},
+		Bootstrap: provision.Bootstrap{
+			X509Provision: provisionX509,
+			Provision:     provisionBS,
+			AutoWhiteList: autoWhiteList,
+			Content:       fmt.Sprintf(mainflux.Env(envBSContent, defBSContent), mainflux.Env(envMQTTURL, defMQTTURL)),
+		},
+
+		// This is default conf for provision if there is no config file
+		Channels: []provision.Channel{
+			{
+				Name:     "control-channel",
+				Metadata: map[string]interface{}{"type": "control"},
+			}, {
+				Name:     "data-channel",
+				Metadata: map[string]interface{}{"type": "data"},
+			},
+		},
+		Things: []provision.Thing{
+			{
+				Name:     "thing",
+				Metadata: map[string]interface{}{"externalID": "xxxxxx"},
+			},
 		},
 	}
-	thingSdkCfg := mfSDK.Config{
-		BaseURL:           mainflux.Env(envThingsLocation, defThingsLocation),
-		HTTPAdapterPrefix: "http",
-		MsgContentType:    "application/json",
-		TLSVerification:   tls,
-	}
-	thingsSDK := mfSDK.NewSDK(thingSdkCfg)
 
-	userSdkCfg := mfSDK.Config{
-		BaseURL:           mainflux.Env(envUsersLocation, defUsersLocation),
-		HTTPAdapterPrefix: "http",
-		MsgContentType:    "application/json",
-		TLSVerification:   tls,
+	cfg.File = mainflux.Env(envConfigFile, "")
+	if cfg.File == "" {
+		cfg.File = defConfigFile
+		if err != provision.Save(cfg, defConfigFile) {
+			return cfg, err
+		}
 	}
-	userSDK := mfSDK.NewSDK(userSdkCfg)
+	return cfg, nil
+}
 
-	cfg.SDK = provsdk.New(cfg.mfCertsURL, cfg.mfBSURL, cfg.mfBSWhiteListURL, thingsSDK, userSDK)
-	return cfg
+func mergeConfigs(dst, src interface{}) interface{} {
```

Comment:
It is really impossible to do something like `func mergeConfigs(src1, src2 interface{}) (dst interface{}) {`?

Reply:
@manuio I think these params explain better what is going on - it is taking one source and merging it into another. If you lable them just as `src1` and `src2`, you will not know which you are mering into which.

I do not know ifthis is importnant, though, but it looks like it is.

Reply:
@manuio I know what you proposed, but this is not good, as the function, in order to produce return value must move one argument into another. How do you know which one is moved into which one, if your argument names do not tell you that?

Reply:
@drasko This could be simplified
```
		mergeConfigs(&cfgFromFile, &cfg)
		cfg = cfgFromFile
```

Reply:
Exactly, destination is mutable, should be passed by reference.